python/tskit/combinatorics.py

```diff
@@ -0,0 +1,723 @@
+
+    @staticmethod
+    def unrank(rank, num_leaves):
+        shape_rank, label_rank = rank
+        unlabelled = RankTree.shape_unrank(shape_rank, num_leaves)
+        return unlabelled.label_unrank(label_rank)
+
+    @staticmethod
+    def shape_unrank(shape_rank, n):
+        """
+        Generate an unlabelled tree with n leaves with a shape corresponding to
+        the `shape_rank`.
+        """
+        part, child_shape_ranks = children_shape_ranks(shape_rank, n)
+        children = [
+            RankTree.shape_unrank(rk, k) for rk, k in zip(child_shape_ranks, part)
+        ]
+
+        t = RankTree(children=children)
+        t._shape_rank = shape_rank
+        return t
+
+    def label_unrank(self, label_rank, labels=None):
+        """
+        Generate a tree with the same shape, whose leaves are labelled
+        from `labels` with the labelling corresponding to `label_rank`.
+        """
+        if labels is None:
+            labels = list(range(self.num_leaves))
+
+        if self.is_leaf():
+            assert label_rank == 0
+            assert len(labels) == 1
+            return RankTree(children=[], label=labels[0])
+
+        child_groups = self.group_children_by_shape()
+        child_labels, child_label_ranks = children_label_ranks(
+            child_groups, label_rank, labels
+        )
+
+        children = self.children
+        labelled_children = [
+            RankTree.label_unrank(c, c_rank, c_labels)
+            for c, c_rank, c_labels in zip(children, child_label_ranks, child_labels)
+        ]
+
+        t = RankTree(children=labelled_children)
+        t._shape_rank = self.shape_rank()
+        t._label_rank = label_rank
+        return t
+
+    @staticmethod
+    def canonical_order(c):
+        """
+        Defines the canonical ordering of sibling subtrees.
+        """
+        return c.num_leaves, c.shape_rank(), c.min_label()
+
+    @staticmethod
+    def from_tsk_tree_node(tree, u):
+        if tree.is_leaf(u):
+            return RankTree(children=[], label=u)
+
+        children = list(
+            sorted(
+                (RankTree.from_tsk_tree_node(tree, c) for c in tree.children(u)),
+                key=RankTree.canonical_order,
+            )
+        )
+        return RankTree(children=children)
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        if len(tree.roots) != 1:
+            raise ValueError("Can't rank trees with multiple roots")
+
+        return RankTree.from_tsk_tree_node(tree, tree.root)
+
+    def to_tsk_tree(self):
+        seq_length = 1
+        tables = tskit.TableCollection(seq_length)
+
+        def add_node(node):
+            if node.is_leaf():
+                assert node.label is not None
+                return node.label
+
+            child_ids = [add_node(child) for child in node.children]
+            # Arbitrarily set parent time +1 from their oldest child
+            max_child_time = max(tables.nodes.time[c] for c in child_ids)
```

Comment:
```suggestion
            max_child_time = max(node_time[c] for c in child_ids)
```

Making tables and arrays is expensive - they're actually all copies at the moment.

Reply:
😮

Reply:
Not that I want to be making copies all the time, but doesn't this mean that `node_time` will remain empty?